browser/ephemeral_storage/ephemeral_storage_tab_helper.cc

```diff
@@ -0,0 +1,183 @@
+/* Copyright (c) 2020 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/ephemeral_storage/ephemeral_storage_tab_helper.h"
+
+#include <set>
+#include <string>
+#include <utility>
+
+#include "base/feature_list.h"
+#include "base/no_destructor.h"
+#include "chrome/browser/profiles/profile.h"
+#include "chrome/browser/ui/browser.h"
+#include "chrome/browser/ui/browser_list.h"
+#include "content/public/browser/browser_context.h"
+#include "content/public/browser/navigation_handle.h"
+#include "content/public/browser/session_storage_namespace.h"
+#include "content/public/browser/storage_partition.h"
+#include "content/public/browser/web_contents.h"
+#include "net/base/registry_controlled_domains/registry_controlled_domain.h"
+#include "third_party/blink/public/common/features.h"
+
+#if defined(OS_ANDROID)
+#include "chrome/browser/android/tab_android.h"
+#include "chrome/browser/ui/android/tab_model/tab_model_list.h"
+#endif
+
+using content::BrowserContext;
+using content::NavigationHandle;
+using content::WebContents;
+
+namespace {
+
+content::SessionStorageNamespaceMap& session_storage_namespace_map() {
+  static base::NoDestructor<content::SessionStorageNamespaceMap>
+      session_storage_namespace_map;
+  return *session_storage_namespace_map.get();
+}
+
+content::SessionStorageNamespaceMap& local_storage_namespace_map() {
+  static base::NoDestructor<content::SessionStorageNamespaceMap>
+      local_storage_namespace_map;
+  return *local_storage_namespace_map.get();
+}
+
+std::string URLToStorageDomain(const GURL& url) {
+  std::string domain = net::registry_controlled_domains::GetDomainAndRegistry(
+      url, net::registry_controlled_domains::INCLUDE_PRIVATE_REGISTRIES);
+
+  // GetDomainAndRegistry might return an empty string if this host is an IP
+  // address or a file URL.
+  if (domain.empty())
+    domain = url::Origin::Create(url.GetOrigin()).Serialize();
```

Comment:
wondering whether it is fully aligned with `security_origin->ToUrlOrigin()` in the renderer. Not a blocker of course, but maybe we could add a test for this

Reply:
I can modify the blink side to avoid calling `SecurityOrigin::ToUrlOrigin()` and to get the `base::Origin` directly from the Document URL. Hopefully that should take care of the possibility of different serialization.

Reply:
I'm not sure whether this is a proper way, since storages are definitely something security-sensitive and AFAIK we should only use SecurityOrigin for sensitive stuff

Reply:
The documentation for url::Origin says it should be suitable for security-relevant decisions:

```c++
// TL;DR: If you need to make a security-relevant decision, use 'url::Origin'.
// If you only need to extract the bits of a URL which are relevant for a
// network connection, use 'url::SchemeHostPort'.
```

That said, this shouldn't override any existing origin-based security. Even if one.a.com and one.b.com have the same ephemeral storage namespace, they won't be able to access each other's DOM storage. The namespace takes care internally of enforcing normal web security protocols.

Reply:
sorry, this is a wrong place for my comment - I was actually concerned with using `window->Url()` in `EphemeralStorageNamespaces::From`. I'm still not sure what is the proper way to do what we want, but I can't provide any evidence that the current way is not secure :)